### ISSUE TITLE
Option to place legend labels near to the data

### ISSUE THREAD

**Issue by Christopher-Bradshaw** [matplotlib/matplotlib#12939]
I was wondering if you had considered a legend option that labels each line near to the line, e.g. from [here](https://www.vis4.net/blog/2012/06/doing-the-line-charts-right/).

![image](https://user-images.githubusercontent.com/3906618/49557052-306fb480-f8bb-11e8-8438-aab778740b76.png)

I was imagining that this could be similar to the current API of:

```
ax.plot(xdata, ydata, label="Oil")
ax.legend(loc="near_line") # or something
```

Automatically deciding where to place the legends might be tricky - mentioned by the [Tufte in R](http://motioninsocial.com/tufte/) package. So there would probably also need to be a way to manually specify locations for tricky cases.

![image](https://user-images.githubusercontent.com/3906618/49557378-45991300-f8bc-11e8-837d-5615a3985008.png)

I know that this can be done manually with `annotate` (this is what I am currently doing, usually just shifted wrt the first/last point of the data) and there is some sample code to somewhat automate it e.g. [1](https://stackoverflow.com/questions/16992038/inline-labels-in-matplotlib), [2](https://github.com/cphyc/matplotlib-label-lines).

Have I missed a native way to do this? If not, is it something that matplotlib would be interested in?

**Comment by jklymak** [comment]
I’d be against this as part of legend, but I’d be in favour as a separate method like label_lines...   Then it could have its own API, preferably close to that of Clabel and annotate. It’d require a bit of bike shedding but could probably be implemented as an experimental feature while we make sure we like the API.

**Comment by QuLogic** [comment]
The [bachelor's degree example](https://matplotlib.org/gallery/showcase/bachelors_degrees_by_gender.html#sphx-glr-gallery-showcase-bachelors-degrees-by-gender-py) does this and could be a basis for any new API.

**Comment by Christopher-Bradshaw** [comment]
Thanks for the pointers to examples + similar methods - having a separate method for this makes sense. If no-one else looks at this, I might take a stab after the holidays. I'll check back in here for other suggestions before starting though.

**Comment by RobertAugustynowicz** [comment]
Hello, our team would like to work on this issue and implement line labels for matplotlib. In thinking of how to implement this feature, we considered making class which inherited artist and contained the necessary information including the legend, as well as the set of texts which would be the labels for each line. But with further examination of your codebase, we see how it could be more favorable to simply implement this in the scripting layer as a method in axes similar to that of annotate as mentioned in the above comment. Would this new way be acceptable, or should we stick to the original way proposed? Any feedback will be appreciated as we would like to make the best possible feature for your project.

**Comment by timhoffm** [comment]
Thanks for the interest! Getting this right is more complex than it looks at first sight. Some random thoughts:

- Where do the annotations live? Are they just free-floating labels or are they grouped in some sort of container. If the latter, can we abuse(?) axis labels for this.
- Related to this: How do we make sure that there is enough space for the labels to display. - Probably solved if we use axis labels. Otherwise, I assume automatic spacing could only be done with constrained_layout.
- Do we want to use the Line2D `label` for the text, or would it be passed as an explicit parameter via the new API?
- How would the API look like? Would it be called per data-line similar to `annotate()` or would it be a global appoach like `legend()`?
- Just providing a new method, which creates/uses existing artists is simpler than defining new types of artists. It's to be preferred if possible. The only down-side would be that the texts are not related to the data points on the code level. Can that be a problem, e.g. with interactive changes of the figure. Do we care and want to handle such cases?

**Comment by RobertAugustynowicz** [comment]
Hello, thank you for the reply. We decided to work with annotations to create and format our text objects in this function. We were considering creating a new container, something as simple as a set of the new annotations so that we can have a record of which text objects belong to the line labels. This would be necessary since we are planning on having the labels redrawn if the graph was resized. To do so we have devised a strategy where we would get the points location in pixels so we can calculate proper spacing based on the font size used in the annotate which can be converted from pt to px. 
In how spacing will be done we have considered two options:
![Case1](https://user-images.githubusercontent.com/25535247/78069717-90b81e80-7368-11ea-9f07-2f8119afcdb0.PNG)
or
![Case2](https://user-images.githubusercontent.com/25535247/78069726-944ba580-7368-11ea-8182-09a46de42734.PNG)
Would  it be preferable to format the labels lines using a bracket like feature when they are too close, or without it?
Also, we have decided for the api to be implemented using a global approach so that all labels are plotted that exist within the legend.
